source/simulator/core.cc

```diff
@@ -1365,6 +1376,92 @@ namespace aspect
 
 
   template <int dim>
+  class VectorFunctionFromVectorFunctionObject : public Function<dim>
+  {
+  public:
+    /**
+     * Given a function object that takes a Point and returns a double
+     * value, convert this into an object that matches the Function@<dim@>
+     * interface.
+     *
+     * @param function_object The scalar function that will form one component
+     *     of the resulting Function object.
+     * @param n_total_components The total number of vector components of the
+     *     resulting Function object.
+     * @param selected_component The single component that should be
+     *     filled by the first argument.
+     **/
+
+    VectorFunctionFromVectorFunctionObject (const std_cxx1x::function<void (const Point<dim> &,Vector<double> &)> &function_object,
+                                            const unsigned int first_component,
+                                            const unsigned int n_object_components,
+                                            const unsigned int n_total_components)
+      :
+      Function<dim>(n_total_components),
+      function_object (function_object),
+      first_component (first_component),
+      n_object_components (n_object_components)
+    {
+//      Assert (n_object_components < this->n_components,
+//              ExcIndexRange (n_object_components, 0, this->n_components));
+    }
+
+    double
+    value (const Point<dim> &p,
+                                                        const unsigned int component) const
+    {
+      Assert (component < this->n_components,
+              ExcIndexRange (component, 0, this->n_components));
+
+      Vector<double> temp(n_object_components);
+      function_object (p, temp);
+
+      if (component < first_component)
+    	  return 0;
+      else if (component >= first_component + n_object_components)
+    	  return 0;
+      else
+        return temp(component - first_component);
```

Comment:
you can move things around to not call the function_object if you are returning 0 anyways.